### PR TITLE
[issue-174] Add Jira Autolink Issue With Comment Test Case

### LINKED ISSUE #174: Add Jira Autolink Issue With Comment Test Case
Add unit test for jira autolink issue with focused comment

### PR DESCRIPTION
**Pull request by sibasankarnayak**
added jira test for url with comment and made it project independent

ticket here 
Fixes #174

### CODE REVIEW

**Review by levb — APPROVED** [review]

**Review by catalintomai — APPROVED** [review]

File: server/autolink/lib_jira_test.go [hunk 1]
```diff
@@ -80,43 +80,75 @@ var jiraTests = []linkTest{
 	}, {
 		"Not relinking",
 		autolink.Autolink{
-			Pattern:  "(MM)(-)(?P<jira_id>\\d+)",
-			Template: "[MM-$jira_id](https://mattermost.atlassian.net/browse/MM-$jira_id)",
+			Pattern:  "(?P<project_id>\\w+)(-)(?P<jira_id>\\d+)",
+			Template: "[${project_id}-${jira_id}](https://mattermost.atlassian.net/browse/${project_id}-${jira_id})",
```
Comment (mickmister) [inline server/autolink/lib_jira_test.go:84]: We aren't using this regex though right? From what I understand about the shorthand links, the Jira plugin will be filling in the project keys, and not use a project-agnostic `pattern`.

I believe this test would make it so something like  


> Here's my phone number-284836364


translates into:

```
Here's my phone [number-284836364](https://mattermost.atlassian.net/browse/number-284836364)
```

2/5 if this undesired behavior is a result of this regex structure, we shouldn't be using it in tests.

cc @hanzei

File: server/autolink/lib_jira_test.go [hunk 2]
```diff
@@ -80,43 +80,75 @@ var jiraTests = []linkTest{
...
+			Pattern:  "(https://mattermost.atlassian.net/browse/)(?P<project_id>\\w+)(-)(?P<jira_id>\\d+)",
+			Template: "[${project_id}-${jira_id}](https://mattermost.atlassian.net/browse/${project_id}-${jira_id})",
 		},
 		"Welcome https://mattermost.atlassian.net/browse/MM-12345",
 		"Welcome [MM-12345](https://mattermost.atlassian.net/browse/MM-12345)",
+	}, {
+		"Comment url replacement",
+		autolink.Autolink{
+			Pattern:  "(https://mattermost.atlassian.net/browse/)(?P<project_id>\\w+)(-)(?P<jira_id>\\d+)[?](focusedCommentId)(=)(?P<comment_id>\\d+)",
+			Template: "[${project_id}-${jira_id} With Comment #${comment_id}](https://mattermost.atlassian.net/browse/${project_id}-${jira_id}?focusedCommentId=${comment_id})",
```
Comment (mickmister) [inline server/autolink/lib_jira_test.go:124]: Not sure if having the comment id in the link text is that helpful. The capitalization of "With Comment" seems weird too.

File: server/autolink/lib_jira_test.go [hunk 3]
```diff
@@ -80,43 +80,75 @@ var jiraTests = []linkTest{
...
+		"Welcome https://mattermost.atlassian.net/browse/MM-12345?focusedCommentId=10210 should link!",
+		"Welcome [MM-12345 With Comment #10210](https://mattermost.atlassian.net/browse/MM-12345?focusedCommentId=10210) should link!",
+	}, {
+		"Comment url replacement multiple times",
+		autolink.Autolink{
+			Pattern:  "(https://mattermost.atlassian.net/browse/)(?P<project_id>\\w+)(-)(?P<jira_id>\\d+)[?](focusedCommentId)(=)(?P<comment_id>\\d+)",
+			Template: "[${project_id}-${jira_id} With Comment #${comment_id}](https://mattermost.atlassian.net/browse/${project_id}-${jira_id}?focusedCommentId=${comment_id})",
+		},
+		"Welcome https://mattermost.atlassian.net/browse/MM-12345?focusedCommentId=10210. should link https://mattermost.atlassian.net/browse/MM-12346?focusedCommentId=10210 !",
+		"Welcome [MM-12345 With Comment #10210](https://mattermost.atlassian.net/browse/MM-12345?focusedCommentId=10210). should link [MM-12346 With Comment #10210](https://mattermost.atlassian.net/browse/MM-12346?focusedCommentId=10210) !",
```
Comment (mickmister) [inline server/autolink/lib_jira_test.go:135]: Any reason why we're testing with a space after the exclamation point? We shouldn't assume that users will put a space there.

**Review by mickmister — COMMENTED** [review]
I have a few questions on this. Since this PR is a proxy for how we will implement this in the Jira plugin, I'd like to resolve concerns here first, even though this is already merged.

**Comment by hanzei** [comment]
Sorry @mich, I should have waited for your review.

File: server/autolink/lib_jira_test.go (hunk 2 above)
Comment (sibasankarnayak) [reply server/autolink/lib_jira_test.go:124]: @mickmister can you suggest a message for this , planning to raise a PR to fix your comments.

File: server/autolink/lib_jira_test.go (hunk 3 above)
Comment (sibasankarnayak) [reply server/autolink/lib_jira_test.go:135]: @mickmister if we don't add space it will still work.

**Comment by sibasankarnayak** [comment]
> I have a few questions on this. Since this PR is a proxy for how we will implement this in the Jira plugin, I'd like to resolve concerns here first, even though this is already merged.

@mickmister sure i raised this PR with your suggested changes , wanted suggestion for message on comment like how should it look like.

This is the PR raised against your suggestion https://github.com/mattermost/mattermost-plugin-autolink/pull/189

File: server/autolink/lib_jira_test.go (hunk 2 above)
Comment (mickmister) [reply server/autolink/lib_jira_test.go:124]: How about:

```
[${project_id}-${jira_id} (comment)](https://mattermost.atlassian.net/browse/${project_id}-${jira_id}?focusedCommentId=${comment_id})
```

which would render as

[MM-1337 (comment)](https://mattermost.atlassian.net/browse/${project_id}-${jira_id}?focusedCommentId=${comment_id})

This is how GitHub does their comment links, such as https://github.com/mattermost/mattermost-plugin-autolink/pull/185#discussion_r841081248

File: server/autolink/lib_jira_test.go (hunk 3 above)
Comment (mickmister) [reply server/autolink/lib_jira_test.go:135]: @sibasankarnayak Right, but the test should reflect the most common use of the software, which I assume would not have a space there.

**Comment by mickmister** [comment]
> Sorry @mickmister, I should have waited for your review.

No worries @hanzei. It took me quite a while to get to this 🙂

**Pull request merged**